docs/notebooks/11-extract-adjoint-solutions.ipynb

```diff
@@ -0,0 +1,2030 @@
+      ],
+      "text/plain": [
+       "<IPython.core.display.HTML object>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "fig, axs = plt.subplots(num_exports+1, 2, sharex='col')\n",
+    "\n",
+    "for i in range(len(forward_solutions)):\n",
+    "    tricontourf(forward_solutions[i], axes=axs[i, 0])\n",
+    "    tricontourf(dFdnu[i], axes=axs[i, 1])\n",
+    "    axs[i, 0].annotate('t={:.2f}'.format(i*timesteps_per_export*dt), (0.05, 0.05), color='white');\n",
+    "    axs[i, 1].annotate('t={:.2f}'.format(i*timesteps_per_export*dt), (0.05, 0.05), color='white');\n",
+    "axs[0, 0].set_title('Forward solution, u');\n",
+    "axs[0, 1].set_title('∂F/∂ν evaluated at u');"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## The objective functional"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In order to solve the adjoint equation, we need an objective functional. For this example, consider the following integral over the 'upstream boundary' at the final time $T$:\n",
+    "\n",
+    "$$J(u)=\\int_{\\Gamma_2}u|_{t=T}\\cdot u_{t=T}\\;\\mathrm ds,\n",
+    "\\quad\\quad\\Gamma_2=\\{(1,y)\\mid y\\in(0,1)\\}.$$"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Objective value: 0.8600\n"
+     ]
+    }
+   ],
+   "source": [
+    "J_form = inner(u, u)*ds(2)\n",
+    "J = assemble(J_form)\n",
+    "print(\"Objective value: {:.4f}\".format(J))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Having solved the forward equation, `firedrake-adjoint` has all it needs to solve the adjoint equation.\n",
+    "Therefore, we can tell it to stop annotating the tape."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "stop_annotating();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Given the definition of $J$, we should exect its derivative w.r.t. $nu$ to be zero:"
```

Comment:
It's a really unfortunate choice of functional if it doesn't depend on the control at all! Could we not use the value of u at the outflow instead?

Reply:
Sorry, I'm not sure I understand. Do you mean we should change the control or the functional? Without realising it, I've used one of the two controls and a very similar functional as in one of the dolfin-adjoint demos: http://www.dolfin-adjoint.org/en/latest/documentation/tutorial.html. The other control used there is the initial condition.

Reply:
Well you've chosen a combination of functional and control such that the functional doesn't depend on the control. That sort of defeats the purpose of solving an adjoint problem! In fact, your functional only depends on the boundary conditions, so it doesn't even really depend on the PDE solution. That's why I suggested changing the functional to be the outflow velocity, because that depends on the PDE solution and therefore on the control. This would result in a nonzero computed gradient.

Reply:
Sorry, I just realised that I wrote 'upstream' in the demo when I meant 'downstream'! The functional is evaluated at the RH boundary, which is the outflow. Is that right, or did you mean something else for the functional?

Reply:
In that case I don't understand the argument that the derivative is expected to be zero.

Reply:
Okay, I've removed the parts with UFL `derivative` calls for simplicity of presentation and added a sentence discussing the negative gradient of J w.r.t. control.